integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go

```diff
@@ -79,57 +81,144 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
-	receiverBytes, err := nodeHandler.GetCoreComponents().AddressPubKeyConverter().Decode(receiver)
+	receiverNonce := uint64(0)
+	receiver2, err := cs.GenerateAndMintWalletAddress(2, chainSim.InitialAmount)
 	require.Nil(t, err)
 
-	receivedTokens := big.NewInt(11)
-	chainSim.TransferESDT(t, cs, wallet.Bytes, receiverBytes, &nonce, tokenToBridge, receivedTokens)
-	chainSim.RequireAccountHasToken(t, cs, tokenToBridge, receiver, receivedTokens)
-	chainSim.RequireAccountHasToken(t, cs, tokenToBridge, wallet.Bech32, big.NewInt(0).Sub(remainingValueAfterBridge, receivedTokens))
+	// will make a cross shard transaction, then deposit transaction, then same shard transaction with every token
+	for _, token := range groupTokens(bridgedInTokens) {
+		amountToSend := big.NewInt(2)
+		if token.Type == core.NonFungibleV2 || token.Type == core.DynamicNFT {
+			amountToSend = big.NewInt(1)
+		}
+		// cross shard transfer
+		receiverToken := transferToken(t, cs, token, wallet, receiver, &nonce, amountToSend)
+		checkTokenDataInAccount(t, cs, token, wallet.Bech32, big.NewInt(0).Sub(token.Amount, receiverToken.Amount))
+		checkTokenDataInAccount(t, cs, token, receiver.Bech32, receiverToken.Amount)
+
+		amountToDeposit := big.NewInt(1)
+		bridgedOutToken := chainSim.ArgsDepositToken{
+			Identifier: token.Identifier,
+			Nonce:      token.Nonce,
+			Amount:     amountToDeposit,
+			Type:       token.Type,
+		}
+
+		// deposit a prefixed token from main chain to sovereign chain,
+		// expecting these tokens to be burned by the whitelisted ESDT safe sc
+		txResult = Deposit(t, cs, receiver.Bytes, &receiverNonce, bridgeData.ESDTSafeAddress, []chainSim.ArgsDepositToken{bridgedOutToken}, receiver.Bytes)
+		chainSim.RequireSuccessfulTransaction(t, txResult)
+
+		remainingTokenAmount := big.NewInt(0).Sub(receiverToken.Amount, amountToDeposit)
+		checkTokenDataInAccount(t, cs, token, receiver.Bech32, remainingTokenAmount)
+		checkTokenDataInAccount(t, cs, token, esdtSafeEncoded, big.NewInt(0))
+		receiverToken.Amount = remainingTokenAmount
+
+		if remainingTokenAmount.Cmp(big.NewInt(0)) > 0 {
+			// same shard transfer
+			transferredToken := transferToken(t, cs, receiverToken, receiver, receiver2, &receiverNonce, remainingTokenAmount)
+			checkTokenDataInAccount(t, cs, transferredToken, receiver.Bech32, big.NewInt(0))
+			checkTokenDataInAccount(t, cs, transferredToken, receiver2.Bech32, remainingTokenAmount)
+		}
+	}
+}
+
+func transferToken(
+	t *testing.T,
+	cs chainSim.ChainSimulator,
+	token chainSim.ArgsDepositToken,
+	sender, receiver dtos.WalletAddress,
+	senderNonce *uint64,
+	amountToSend *big.Int,
+) chainSim.ArgsDepositToken {
+	if token.Type == core.Fungible {
+		chainSim.TransferESDT(t, cs, sender.Bytes, receiver.Bytes, senderNonce, token.Identifier, amountToSend)
+	} else {
+		chainSim.TransferESDTNFT(t, cs, sender.Bytes, receiver.Bytes, senderNonce, token.Identifier, token.Nonce, amountToSend)
+		err := cs.GenerateBlocks(3)
```

Comment:
Why generate extra blocks here? 
Because cross shard txs? And why it works for `TransferESDT` without generated blocks?

Reply:
For NFT transfer, sender == receiver in arguments and actual receiver is in input data in tx. So the chain simulator verifies the status on the same shard and is successful in 1 block. Then I need to wait 3 more blocks so the actual receiver will receive the tokens because is in different shard.